Add support for excluding boundary ids from sideset generation in a number of sideset-generating MGs

GiudGiud opened the pull request:
closes #22117 https://github.com/idaholab/moose/issues/28551

GiudGiud committed:
Rename some attributes of sidesetgeneratorbase for simplicity
Add support for excluding boundary ids from sideset generation in a number of sideset-generating MGs

GiudGiud committed:
Add more bad input checking to parsed generate sideset refs #22117

GiudGiud committed:
Add test for excluding boundaries from sideset generation
refs #28551

GiudGiud committed:
Add checks for bad subdomain inputs in advanced extruder, refs #22117

GiudGiud committed:
Add checks for bad subdomain inputs in lower D block from sideset gen, refs #22117

GiudGiud committed:
Typo fixes in MGs

GiudGiud committed:
Add checks for bad subdomain inputs in XY line cutter, closes #22117

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2428164/) on e3c4820 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/28550/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2428179/) on e3c4820 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>9a6b27</th>
      <th colspan=4>#28550 e3c482</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.03%</td>
      <td>85.02%</td>
      <td>-0.01%</td>
      <td>75.58%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>105242</td>
      <td>105296</td>
      <td>+54</td>
      <td>65</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>18528</td>
      <td>18546</td>
      <td>+18</td>
      <td>21</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/28550/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/28550/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/28550/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/28550/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/28550/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/28550/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/28550/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/28550/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/28550/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/28550/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/28550/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/28550/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/28550/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/28550/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/28550/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/28550/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/28550/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/28550/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/28550/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/28550/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/28550/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/28550/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/28550/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/28550/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/28550/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/28550/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/28550/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/28550/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/28550/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/28550/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 75.58% is less than the suggested 90.0%



This comment will be updated on new commits.

pbehne commented:
I've started to review this. Should be able to finish by next week.

GiudGiud commented:
Awesome thanks!

pbehne commented on framework/src/meshgenerators/SideSetsGeneratorBase.C:
```diff
@@ -40,10 +40,16 @@ SideSetsGeneratorBase::validParams()
+      "A set of boundary names or ids whose sides will be included in the new sidesets.  A side "
+      "is only added if it also belongs to one of these boundaries.");
+  params.addParam<std::vector<BoundaryName>>(
+      "excluded_boundaries",
```
Add param name to group on line 77.

pbehne reviewed (changes requested):
Please reference issue 22117 in the PR description.

Do these changes fully address 22117? Or are there other mesh generators that also need to be updated? i.e., should 22117 be closed, or left open?

If these changes fully address 22117, what process did you use to verify that you caught all mesh generators that need block/sideset checking added?

GiudGiud commented on framework/src/meshgenerators/SideSetsGeneratorBase.C:
```diff
@@ -251,7 +295,18 @@ SideSetsGeneratorBase::elementSideInIncludedBoundaries(const Elem * const elem,
                                                        const MeshBase & mesh) const
 {
-  for (const auto bid : _restricted_boundary_ids)
+  for (const auto bid : _included_boundary_ids)
```
no need to take references of trivially copyable objects. it s actually slower

GiudGiud commented on framework/src/meshgenerators/AdvancedExtruderGenerator.C:
```diff
@@ -177,7 +177,9 @@ AdvancedExtruderGenerator::AdvancedExtruderGenerator(const InputParameters & par
 
   if (_subdomain_swaps.size() && (_subdomain_swaps.size() != num_elevations))
     paramError("subdomain_swaps",
-               "If specified, 'subdomain_swaps' must be the same length as 'heights' in ",
+               "If specified, 'subdomain_swaps' (" + std::to_string(_subdomain_swaps.size()) +
+                   ") must be the same length as 'heights' (" + std::to_string(num_elevations) +
+                   ") in ",
```
no there are not. there are a lot of missing exception checks

GiudGiud commented on framework/src/meshgenerators/LowerDBlockFromSidesetGenerator.C:
```diff
@@ -85,6 +85,11 @@ LowerDBlockFromSidesetGenerator::generate()
   auto sideset_ids = MooseMeshUtils::getBoundaryIDs(*mesh, _sideset_names, true);
   std::set<boundary_id_type> sidesets(sideset_ids.begin(), sideset_ids.end());
 
+  // Check that the sidesets are present in the mesh
+  for (const auto & sideset : _sideset_names)
+    if (!MooseMeshUtils::hasBoundaryName(*mesh, sideset))
+      paramError("sidesets", "The sideset '", sideset, "' was not found within the mesh");
+
```
yeah that's the problem

GiudGiud commented on framework/src/meshgenerators/XYMeshLineCutter.C:
```diff
@@ -110,6 +110,18 @@ XYMeshLineCutter::generate()
       *(replicated_mesh_ptr->elem_dimensions().rbegin()) != 2)
     paramError("input", "Only 2D meshes are supported.");
 
+  // Check that the input boundaries are part of the mesh
+  if (isParamValid("input_mesh_external_boundary_id"))
+    if (!MooseMeshUtils::hasBoundaryID(*replicated_mesh_ptr, _input_mesh_external_boundary_id))
+      paramError("input_mesh_external_boundary_id", "Boundary must exist in input mesh");
+
+  // Check that the other boundaries to conform to are part of the mesh
+  if (isParamValid("other_boundaries_to_conform"))
+    for (const auto bid : _other_boundaries_to_conform)
+      if (!MooseMeshUtils::hasBoundaryID(*replicated_mesh_ptr, bid))
+        paramError("other_boundaries_to_conform",
+                   "Boundary '" + std::to_string(bid) + "' must exist in input mesh");
+
```
you need a pointer to the mesh for them

GiudGiud committed:
Address Patrick's review

- re organize checks
- improve docstrings & parameters
- use num_elevations instead of re-computing it

Co-authored-by: Patrick Behne <31858053+pbehne@users.noreply.github.com>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2448925/) on 49c538f wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/28550/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/28550/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format d13cb8d6e6685e7ac1e929193fee033e139b8379`

GiudGiud commented on framework/src/meshgenerators/SideSetsGeneratorBase.C:
```diff
@@ -135,17 +143,53 @@ SideSetsGeneratorBase::setup(MeshBase & mesh)
                    "' was not found within the mesh");
   }
 
+  if (_check_excluded_boundaries)
+  {
+    const auto & excluded_boundaries = getParam<std::vector<BoundaryName>>("excluded_boundaries");
+    for (const auto & boundary_name : _boundary_names)
+      if (std::find(excluded_boundaries.begin(), excluded_boundaries.end(), boundary_name) !=
+          excluded_boundaries.end())
+        paramError(
+            "new_boundary",
+            "A boundary cannot be both the new boundary and be excluded in the list of excluded "
+            "boundaries.");
+    _excluded_boundary_ids = MooseMeshUtils::getBoundaryIDs(mesh, excluded_boundaries, false);
+
+    // Check that the excluded boundary ids/names exist in the mesh
+    for (const auto & i : index_range(_excluded_boundary_ids))
+      if (_excluded_boundary_ids[i] == Moose::INVALID_BOUNDARY_ID)
+        paramError("excluded_boundaries",
+                   "The boundary '",
+                   excluded_boundaries[i],
+                   "' was not found within the mesh");
```
It should not.
I dont think it will be more readable to have a lambda helper

GiudGiud commented on framework/src/meshgenerators/SideSetsGeneratorBase.C:
```diff
@@ -251,7 +295,18 @@ SideSetsGeneratorBase::elementSideInIncludedBoundaries(const Elem * const elem,
+}
+
+bool
+SideSetsGeneratorBase::elementSideInExcludedBoundaries(const Elem * const elem,
```
I would argue it's too trivial to be moved in MooseMeshUtils.

saving 3 lines of code by creating a helper with 5 lines in header (docstring) and 5 other in the source does not help readability imo

GiudGiud commented on test/tests/meshgenerators/xy_mesh_line_cutter/tests:
```diff
@@ -63,73 +63,92 @@
-    requirement = 'The system shall throw an error if the external boundary id is not provided when the node moving method is used.'
-    cli_args = 'Mesh/mlc/cutting_type=MOV_NODE
-                --mesh-only simple_cut_err.e'
+  [errors]
```
so for nested tests it's actually quite annoying to add I think. i just tried something, we ll see what SQA shows

GiudGiud commented on framework/src/meshgenerators/SideSetsGeneratorBase.C:
```diff
@@ -135,17 +143,53 @@ SideSetsGeneratorBase::setup(MeshBase & mesh)
+            _excluded_boundary_ids.end())
+          paramError("excluded_boundaries",
+                     "'included_boundaries' and 'excluded_boundaries' lists should not overlap");
+      for (const auto & boundary_id : _excluded_boundary_ids)
+        if (std::find(_included_boundary_ids.begin(), _included_boundary_ids.end(), boundary_id) !=
+            _included_boundary_ids.end())
+          paramError("excluded_boundaries",
+                     "'included_boundaries' and 'excluded_boundaries' lists should not overlap");
```
you're right. I was thinking about entries in one but not the other when I doubled it up, but obviously then there is no overlap

GiudGiud commented on framework/src/meshgenerators/SideSetsGeneratorBase.C:
```diff
@@ -135,17 +143,53 @@ SideSetsGeneratorBase::setup(MeshBase & mesh)
+    for (const auto & boundary_name : _boundary_names)
+      if (std::find(excluded_boundaries.begin(), excluded_boundaries.end(), boundary_name) !=
+          excluded_boundaries.end())
+        paramError(
+            "new_boundary",
+            "A boundary cannot be both the new boundary and be excluded in the list of excluded "
+            "boundaries.");
```
added

GiudGiud commented on framework/src/meshgenerators/SideSetsGeneratorBase.C:
```diff
@@ -135,17 +143,53 @@ SideSetsGeneratorBase::setup(MeshBase & mesh)
+        paramError("included_boundaries",
                    "The boundary '",
                    included_boundaries[i],
                    "' was not found within the mesh");
```
reverted, this does not report the user-selected name, it outputs the invalid id instead which does not help the user

pbehne reviewed (approved):
The following question from my first review has not been addressed:

> If these changes fully address 22117, what process did you use to verify that you caught all mesh generators that need block/sideset checking added?

For instance, are there any mesh generators in the `reactor` module that take blocks and sidesets as input parameters and fails to check if they exist?

In any case, I think quickly merging these useful changes takes precedence over finding EVERY applicable mesh generator (unless someone REALLY wants to do this). These can always be fixed as they are found.

In the future, would it be possible/useful to define new block/sideset types that are tied to a specific mesh? For instance, we currently have types `BoundaryName` and `SubdomainName`. What if we define new types `ExistingBoundaryName` and `ExistingSubdomainName` that have to be tied to a particular mesh upon instantiation? Then the error checks could simply be built into these new types and all that would need to be updated everywhere in these mesh generators is the parameter type. I'me sure it gets more complicated than this, though (i.e., rechecking that blocks and boundaries still exists every time the mesh changes...).

GiudGiud commented:
> For instance, are there any mesh generators in the reactor module that take blocks and sidesets as input parameters and fails to check if they exist?

I dont think so. We have been more careful from the very start in the reactor module. Yinbin did a phenomenal job at adding checks and covering them in the test suite too. 

> In any case, I think quickly merging these useful changes takes precedence over finding EVERY applicable mesh generator (unless someone REALLY wants to do this). These can always be fixed as they are found.

Checking every single MG is what I did for the framework actually. I can do it for the reactor module at some point.

> What if we define new types ExistingBoundaryName and ExistingSubdomainName that have to be tied to a particular mesh upon instantiation

Could be a solution but this feels like it would need to be handled near parameter parsing, when really it's being handled way later in mesh generation.
Maybe adding getSubdomainID (....., bool must_exist) is the simpler solution, but as you mentioned this makes us lose context

GiudGiud commented:
@roystgnr

roystgnr reviewed (approved):
Should we get something about these parameters into its own `framework/doc` page?  It seemed to me that the wording of the parameters doc strings implied that the effect of specifying both lists should be a logical-and, but I wasn't really confident of that until reading the source code.

GiudGiud commented:
We ll need to think about how to do this. the capability is in sideset generator base so multiple MGs could use the doc. I dont really see users look at base class doco, so maybe just duplicate the doco for each

roystgnr commented:
Looks like we can use a `!include` directive for the duplication?  I'm seeing that in a few `.md` files at least.
